### PR TITLE
:sparkles: Add Test-Time Adaptation

### PR DESCRIPTION
**Pull request by o-laurent**
Remains to be done:
- [x] Agree on the structure
- [ ] Make the documentation
- [ ] Add a tutorial on using any TTA method
- [x] Add @FarinaMatteo's Zero 
  - [x] Add Zero to the references
- [ ] ~Add a tutorial on evaluating Zero with CLIP~
  - [ ] ~Install openclip on chocolatine~
- [ ] Check that `shuffle` is `False` when `num_tta >1`
- [ ] Fill the coverage

### CODE REVIEW

File: torch_uncertainty/models/wrappers/zero.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+        Zero performs "0-temperature averaging" (i.e. majority voting) at evaluation. It starts
+        by filtering the :attr:`filter_views` most confident predictions, and returns the majority vote
+        as a prediction. If used during training, the predictions will be those of the inner-model
+        passed as argument (:attr:`model`).
+
+        Args:
+            model (nn.Module): The inner model to train.
+            num_tta (int): The number of views at evaluation time.
+            filter_views (float): Filter out 1-:attr:`filter_views` of the predictions of the augmented views.
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply torch_uncertainty/models/wrappers/zero.py:19]: I agree, but it's not a percent, so I didn't know how to frame it

File: torch_uncertainty/models/wrappers/zero.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        batch_size, _, num_classes = all_predictions.shape
+        entropies = entr(all_predictions).sum(1)
+
+        # Get the index of the most confident predictions on the views
+        conf_idx = torch.argsort(entropies, dim=-1)
+        votes = all_predictions.argmax(-1)
+
+        # Count the votes
+        predictions = torch.zeros((batch_size, num_classes), device=all_predictions.device)
+        for img_id, img_votes in enumerate(votes):
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply torch_uncertainty/models/wrappers/zero.py:40]: bincount only works with 1d tensors, so I don't think that we can avoid the outer loop. And the inner-loop is mandatory

File: torch_uncertainty/models/wrappers/zero.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (o-laurent) [reply torch_uncertainty/models/wrappers/zero.py:19]: Maybe share?

File: torch_uncertainty/models/wrappers/zero.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (o-laurent) [reply torch_uncertainty/models/wrappers/zero.py:40]: We could use the original softmax(x/eps) but it's behaviour may be undefined (we lose some important properties of the softmax at such magnitudes)

**Review by alafage — APPROVED** [review]
Thanks @o-laurent ! Looks good on my side
